volume/devicemapper/devicemapper.go

```diff
@@ -1135,86 +1236,144 @@ func (d *DeviceMapperDriver) GetTenantStorageStats() ([]volume.TenantStorageStat
+			// valid driver.
+			bfr := &bytes.Buffer{}
+			if _, err := bfr.ReadFrom(tarfile); err != nil {
+				glog.Errorf("Could not read from %s: %s", driverFile, err)
+				return err
+			}
+			driverType = bfr.String()
+			glog.V(2).Infof("Volume driver type is %s", driverType)
+		} else if header.Name == deviceFile {
+
+			// Update the staging device to match the provided settings
+			volInfo := volumeInfo{}
+			if err := json.NewDecoder(tarfile).Decode(&volInfo); err != nil {
+				glog.Errorf("Could not read from %s: %s", deviceFile, err)
+				return err
+			}
+
+			// Ensure the size of the device is greater than or equal to the
+			// value specified in the settings.
+			size, err := v.driver.deviceSize(deviceID)
+			if err != nil {
+				glog.Errorf("Could not determine the size of the staging device %s: %s", deviceID, err)
 				return err
 			}
-			drivertype = buf.String()
-		} else if strings.HasPrefix(header.Name, volumedir) {
-			header.Name = strings.Replace(header.Name, volumedir, label, 1)
+			if volInfo.Size > size {
+				glog.Warning("Device size from import %s is greater than size of staging device; expanding.", label, deviceID)
+				if err := v.driver.resize(deviceID, volInfo.Size); err != nil {
+					glog.Errorf("Could not resize device %s; not importing snapshot %s: %s", deviceID, label, err)
+					return err
+				}
+				glog.V(2).Infof("Device %s is now %s", deviceID, units.HumanSize(float64(volInfo.Size)))
+			}
+		} else if strings.HasPrefix(header.Name, volumeDir) {
+
+			// Untar into mountpoint
+			header.Name = strings.TrimPrefix(header.Name, volumeDir)
 			if err := volume.ImportArchiveHeader(header, tarfile, mountpoint); err != nil {
 				return err
 			}
-		} else if strings.HasPrefix(header.Name, metadatadir) {
-			header.Name = strings.Replace(header.Name, metadatadir, label, 1)
-			if err := volume.ImportArchiveHeader(header, tarfile, v.driver.MetadataDir()); err != nil {
+		} else if strings.HasPrefix(header.Name, metaDir) {
+
+			// Untar into the metadata
+			header.Name = strings.TrimPrefix(header.Name, metaDir)
+			if err := volume.ImportArchiveHeader(header, tarfile, metaPath); err != nil {
 				return err
 			}
 		}
 	}
-	if drivertype == "" {
-		return errors.New("incompatible snapshot")
+
+	// Add device as a snapshot of this volume.
+	if err := v.Metadata.AddSnapshot(label, deviceID); err != nil {
+		glog.Errorf("Could not save device %s as snapshot %s: %s", deviceID, label, err)
+		return err
 	}
-	return v.Metadata.AddSnapshot(label, device)
+
+	// Set the snapshot metadata path
+	if err := os.Rename(metaPath, filepath.Join(filepath.Dir(metaPath), label)); err != nil {
+		glog.Errorf("Could not set device metadata for snapshot %s: %s", label, err)
+		return err
+	}
+
+	return nil
+}
+
+func (v *DeviceMapperVolume) SizeOf() (uint64, error) {
+	return v.driver.deviceSize(v.deviceID())
 }
 
-func (v *DeviceMapperVolume) volumeDevice() string {
+func (v *DeviceMapperVolume) deviceID() string {
```

Comment:
Why'd you rename this? deviceID is somewhat more ambiguous, since there are many devices associated with a volume (the volume device itself, and the snapshot devices). Maybe tenantDeviceID? or currentDeviceID? or whatever